ServerCore/Pages/Puzzles/SinglePlayerPuzzleHints.cshtml.cs

```diff
@@ -91,12 +91,36 @@ public async Task<IActionResult> OnPostUnlockAsync(int hintID, int puzzleId)
 
         private async Task<IList<HintWithState>> GetHints(int puzzleID)
         {
-            Hints = await (from Hint hint in _context.Hints
+            List<HintWithNullableState> hintsWithNullableStates = await (from Hint hint in _context.Hints
                            where hint.Puzzle.ID == puzzleID
-                           join SinglePlayerPuzzleHintStatePerPlayer state in _context.SinglePlayerPuzzleHintStatePerPlayer on hint.Id equals state.HintID
-                           where state.PlayerID == LoggedInUser.ID
+                           join SinglePlayerPuzzleHintStatePerPlayer state in _context.SinglePlayerPuzzleHintStatePerPlayer on hint.Id equals state.HintID into leftJoinedTable
```

Comment:
I've never written a left join and would be happier if @morganbr signs off on this part. The rest looks good though!

Reply:
This looks ok, but usually we eagerly create state when a puzzle or player gets created rather than needing to lazily fix it up later like this